observability/src/main/java/io/grpc/observability/interceptors/GcpLogHelper.java

```diff
@@ -0,0 +1,430 @@
+import io.grpc.observability.logging.Sink;
+import io.grpc.observabilitylog.v1.GrpcLogRecord;
+import io.grpc.observabilitylog.v1.GrpcLogRecord.Address;
+import io.grpc.observabilitylog.v1.GrpcLogRecord.EventType;
+import io.grpc.observabilitylog.v1.GrpcLogRecord.LogLevel;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.Inet4Address;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * Helper class for GCP observability logging.
+ */
+final class GcpLogHelper {
+  private static final Logger logger = Logger.getLogger(GcpLogHelper.class.getName());
+  static final Metadata.Key<byte[]> STATUS_DETAILS_KEY =
+      Metadata.Key.of(
+          "grpc-status-details-bin",
+          Metadata.BINARY_BYTE_MARSHALLER);
+
+  /**
+   *  Class for proto related helper methods.
+   */
+  static final class GcpLogSinkWriter {
+    private final Sink sink;
+    private final TimeProvider timeProvider;
+
+    GcpLogSinkWriter(Sink sink, TimeProvider timeProvider) {
+      this.sink = sink;
+      this.timeProvider = timeProvider;
+    }
+
+    GrpcLogRecord.Builder createTimestamp() {
+      long nanos = timeProvider.currentTimeNanos();
+      return GrpcLogRecord.newBuilder().setTimestamp(Timestamps.fromNanos(nanos));
+    }
+
+
+    /**
+     * Logs the client header.
+     */
+    void logClientHeader(
+        long seqId,
+        String serviceName,
+        String methodName,
+        @Nullable String authority,
+        @Nullable Duration timeout,
+        Metadata metadata,
+        GrpcLogRecord.EventLogger eventLogger,
+        String rpcId,
+        @Nullable SocketAddress peerAddress) {
+      checkNotNull(serviceName, "serviceName");
+      checkNotNull(methodName, "methodName");
+      checkNotNull(rpcId, "rpcId");
+      checkArgument(
+          peerAddress == null || eventLogger == GrpcLogRecord.EventLogger.LOGGER_SERVER,
+          "peerAddress can only be specified by server");
+
+      PayloadBuilder<GrpcLogRecord.Metadata.Builder> pair = createMetadataProto(metadata);
+      GrpcLogRecord.Builder logEntryBuilder = createTimestamp()
+          .setSequenceId(seqId)
+          .setServiceName(serviceName)
+          .setMethodName(methodName)
+          .setEventType(EventType.GRPC_CALL_REQUEST_HEADER)
+          .setEventLogger(eventLogger)
+          .setLogLevel(LogLevel.LOG_LEVEL_DEBUG)
+          .setMetadata(pair.proto)
+          .setPayloadSize(pair.size)
+          .setRpcId(rpcId);
+      if (authority != null) {
+        logEntryBuilder.setAuthority(authority);
+      }
+      if (timeout != null) {
+        logEntryBuilder.setTimeout(timeout);
+      }
+      if (peerAddress != null) {
```

Comment:
How about encapsulating the boilerplate stuff in specific methods? So your `createTimestamp()` creates the `logEntryBuilder` and you then call setter on them. Instead you can define - say - `forClientHeader(authority, timeout, peerAddress)` which is something like:

```
GrpcLogRecord.Builder forClientHeader(String authority,
        @Nullable Duration timeout,
        @Nullable SocketAddress peerAddress) {

      GrpcLogRecord.Builder logEntryBuilder = createTimestamp();
      if (authority != null) {
        logEntryBuilder.setAuthority(authority);
      }
      if (timeout != null) {
        logEntryBuilder.setTimeout(timeout);
      }
      if (peerAddress != null) {
        logEntryBuilder.setPeerAddress(socketToProto(peerAddress));
      }
     return logEntryBuilder;
}
```

So the code in `logClientHeader()` is much simpler. Create similar ones for `forTrailer` `forServerHeader` and so on. Only do it if you agree

Reply:
For now, with authority being non-null and similarly for peerAddress being non-null on server(once we know the reason behind Binary Logging's implementation). We do not have many non-null checks anymore. So I kept the structure as is, if thats not convincing please lmk, I will make the changes.